core/src/providers/google_vertex_ai.rs

```diff
@@ -0,0 +1,527 @@
+                    if completion.candidates.len() != 1 {
+                        Err(anyhow!(
+                            "Unexpected number of candidates: {}",
+                            completion.candidates.len()
+                        ))?;
+                    }
+                    if completion.candidates[0].content.parts.len() != 1 {
+                        Err(anyhow!(
+                            "Unexpected number of parts: {}",
+                            completion.candidates[0].content.parts.len()
+                        ))?;
+                    }
+
+                    match event_sender.as_ref() {
+                        Some(sender) => {
+                            let text = completion.candidates[0].content.parts[0].text.clone();
+                            if text.len() > 0 {
+                                let _ = sender.send(json!({
+                                    "type": "tokens",
+                                    "content": {
+                                        "text": text,
+                                    }
+                                }));
+                            }
+                        }
+                        _ => (),
+                    }
+
+                    completions.lock().push(completion);
+                }
+                None => {
+                    break 'stream;
+                }
+            },
+            Err(e) => match e {
+                es::Error::Eof => break 'stream,
+                _ => {
+                    Err(anyhow!("Error streaming tokens from Vertex AI: {:?}", e))?;
+                    break 'stream;
+                }
+            },
+        }
+    }
+
+    let mut completion = Completion {
+        candidates: vec![Candidate {
+            content: Content {
+                role: String::from("MODEL"),
+                parts: vec![Part {
+                    text: String::from(""),
+                }],
+            },
+            finish_reason: None,
+        }],
+        usage_metadata: Some(UsageMetadata {
+            prompt_token_count: 0,
+            candidates_token_count: 0,
+            total_token_count: 0,
+        }),
+    };
+
+    completions.lock().iter().for_each(|c| {
+        completion.candidates[0].content.parts[0].text.push_str(
+            c.candidates[0]
+                .content
+                .parts
+                .iter()
+                .map(|p| p.text.as_str())
+                .collect::<Vec<&str>>()
+                .join(" ")
+                .as_str(),
+        );
+        if c.candidates[0].finish_reason.is_some() {
+            completion.candidates[0].finish_reason = c.candidates[0].finish_reason.clone();
+        }
+        if c.usage_metadata.is_some() {
+            completion.usage_metadata = c.usage_metadata.clone();
+        }
+    });
+
+    Ok(completion)
+}
+
+pub async fn get_access_token(service_account_json: &str) -> Result<String> {
```

Comment:
```suggestion
async fn get_access_token(service_account_json: &str) -> Result<String> {
```

Since it's not used externally.

Reply:
Moved that to the class